### ISSUE TITLE
New sink: postgres

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#15765]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Users want to write data to Postgres from Vector.

Breaking this off from https://github.com/vectordotdev/vector/issues/939

### References

- https://github.com/vectordotdev/vector/issues/6556
- https://github.com/vectordotdev/vector/issues/939

### Version

vector 0.26.0

**Comment by hhromic** [comment]
This is certainly an interesting kind of sink: databases.
I wonder if it would make more sense to implement an ODBC sink instead of specific database sinks?
This way, any database that provides ODBC drivers (there are many already) could interface with Vector out of the box. In addition, probably would be much less maintenance burden for Vector than maintaining multiple specific sinks.

**Comment by jszwedko** [comment]
Agreed, supporting something like ODBC would let us hit a lot of them at once. https://github.com/pacman82/odbc-api seems promising as a Rust wrapper.

**Comment by zamazan4ik** [comment]
AFAIK, rsyslog implements both approaches: ODBC based and database-specific based. The first one allows using with the same plugin for as many databases as possible without database-specific features. The second one uses database-specific features.

**Comment by spencergilbert** [comment]
I was also going to suggest a more general sink, at least to start with

**Comment by thomasdba** [comment]
any updates ?

**Comment by vnagendra** [comment]
Unless I am missing something what is needed for Postgres specifically is already in the current `Cargo.toml` (according to the PRs on this repo anyway).. 

https://github.com/vectordotdev/vector/pull/18391/files

My suggestion would be to have something that addresses a specific set of (potentially limited) use cases. For me, having the ability to write to Db (a specific table) avoids having 1 more thing in the middle like Redis/Rabbit/etc., 

This is a fair thing to do for a limited number of things received (logs, metrics, messages whatever). When I say limited I am talking 100K+/day -- that is enough for a good set of use cases. Hopefully nobody receiving 100K/min is going to write to a db :)

**Comment by jorgehermo9** [comment]
I would like to give a try to this. It should be similar to `databend` and `clickhouse` sinks, right?

I see that we could encode the event as json (like the `clickhouse` sink does) and then use `populate_json_record` function from postgres ([related docs](https://www.postgresql.org/docs/current/functions-json.html)) to fill the table schema from the json.

**Comment by jszwedko** [comment]
> I would like to give a try to this, if I have enough time these days, I could submit a PR addressing this in a couple of weeks. It should be similar to `databend` and `clickhouse` sinks, right?
> 
> I see that we could encode the event as json (like the `clickhouse` sink does) and then use `populate_json_record` function from postgres ([related docs](https://www.postgresql.org/docs/current/functions-json.html)) to fill the table schema from the json.

Awesome! It seems like there is significant demand for this feature. I think what you are describing could work. I also think we could leverage https://github.com/apache/opendal to do the inserts. We started using this library with the `webhdfs` sink as a way to try it out and have a PR using it for an `sftp` sink: https://github.com/vectordotdev/vector/pull/18076. It seems to be a promising approach to expand Vector's sinks without minimal wrapping.

**Comment by jorgehermo9** [comment]
It seems that OpenDal uses postgres as a key-value storage

https://github.com/apache/opendal/blob/main/core/src/services/postgresql/backend.rs#L187

**Comment by jorgehermo9** [comment]
Also, it is worth noting this statement from [Opendal's documentation](https://opendal.apache.org/docs/overview/#what-does-opendal-do)
![image](https://github.com/user-attachments/assets/93a9789f-f919-4fe6-8eab-f6b557dec4d8)

**Comment by jszwedko** [comment]
> Also, it is worth noting this statement from [Opendal's documentation](https://opendal.apache.org/docs/overview/#what-does-opendal-do) ![image](https://private-user-images.githubusercontent.com/23034060/364139003-93a9789f-f919-4fe6-8eab-f6b557dec4d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0MDA5NjEsIm5iZiI6MTcyNTQwMDY2MSwicGF0aCI6Ii8yMzAzNDA2MC8zNjQxMzkwMDMtOTNhOTc4OWYtZjkxOS00ZmU2LThlYWItZjZiNTU3ZGVjNGQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAzVDIxNTc0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0ZTE1NjY3MDc2OWIxNWUwNzNhNzg0M2JkZTNmYmJmYzdmMWQ0ZWFjMzk1NjhlZGU5MTE0MzRhODM0ZDY5MWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.h-UYblaJsxbAHu9bcGIgeVn_T2rU9dxCDn7newGVPBE)

Ah, good find 👍 I think what you suggested with `json_populate_recordset` likely makes the most sense then.

**Comment by jorgehermo9** [comment]
CC @Ichmed @isbm 🎉

**Comment by isbm** [comment]
@jorgehermo9 wonderful! Thanks for your hard work. Now, invent something else cool. We will wait a bit, and then troll you with our own PR. 😛 (joking)

Thanks a lot, again!
